File: packages/snow_draw_engine/lib/draw/elements/types/rectangle/rectangle_task_encoder.dart
import '../../../models/element_state.dart';
import '../../../render/tasks/render_tasks.dart';
import '../../../services/text/text_metrics_service.dart';
import '../../core/typed_element_render_task_encoder.dart';
import 'rectangle_data.dart';

/// Encodes rectangles into high-level render tasks.
final class RectangleTaskEncoder
    extends TypedElementRenderTaskEncoder<RectangleData> {
  /// Creates a rectangle task encoder.
  const RectangleTaskEncoder();

  @override
  List<RenderTask> encodeTypedTasks({
    required ElementState element,
    required RectangleData data,
    String? localeTag,
    TextMetricsService? textMetricsService,
  }) => <RenderTask>[
    RectangleRenderTask(element: element, data: data, localeTag: localeTag),
  ];
}

File: packages/snow_draw_rust_engine/src/draw/elements/types/rectangle/rectangle_task_encoder.rs
#![allow(dead_code)]

use std::sync::Arc;

use crate::draw::elements::core::typed_element_render_task_encoder::{
    ElementState, RenderTaskList, TypedElementRenderTaskEncoder,
};
use crate::draw::types::edit_context::TextMetricsService;

use super::rectangle_data::RectangleData;

/// Element-level render task for rectangle payloads.
///
/// Mirrors Dart `RectangleRenderTask(element: ..., data: ..., localeTag: ...)`.
#[derive(Debug, Clone)]
pub struct RectangleRenderTask {
    pub element: ElementState,
    pub data: RectangleData,
    pub locale_tag: Option<String>,
}

impl RectangleRenderTask {
    /// Creates a new rectangle render task snapshot.
    pub fn new(element: ElementState, data: RectangleData, locale_tag: Option<String>) -> Self {
        Self {
            element,
            data,
            locale_tag,
        }
    }
}

/// Encodes rectangles into high-level render tasks.
///
/// Translation of Dart
/// `RectangleTaskEncoder extends TypedElementRenderTaskEncoder<RectangleData>`.
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub struct RectangleTaskEncoder;

impl RectangleTaskEncoder {
    /// Creates a rectangle task encoder.
    pub const fn new() -> Self {
        Self
    }
}

impl TypedElementRenderTaskEncoder<RectangleData> for RectangleTaskEncoder {
    fn encode_typed_tasks(
        &self,
        element: &ElementState,
        data: &RectangleData,
        locale_tag: Option<&str>,
        _text_metrics_service: Option<Arc<dyn TextMetricsService>>,
    ) -> RenderTaskList {
        vec![Box::new(RectangleRenderTask::new(
            element.clone(),
            data.clone(),
            locale_tag.map(str::to_owned),
        ))]
    }
}
